api/src/main/java/org/eclipse/microprofile/reactive/messaging/connector/OutgoingConnectorFactory.java

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright (c) 2018-2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.eclipse.microprofile.reactive.messaging.connector;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.MessagingProvider;
+import org.eclipse.microprofile.reactive.streams.operators.SubscriberBuilder;
+
+import java.util.NoSuchElementException;
+
+/**
+ * SPI used to implement a connector managing a sink of messages for a specific <em>transport</em>. Typically, to
+ * handle dispatch messages to Kafka, the reactive messaging extension would need to implement a {@code bean}
+ * implementing this interface. This bean is called for every {@code stream} that need to be created for this specific
+ * <em>transport</em> (so Kafka in the example). These streams are connected to methods annotated with
+ * {@link org.eclipse.microprofile.reactive.messaging.Outgoing}.
+ * <p>
+ * The factory is called to create a {@code subscriber} for each configured <em>transport</em>. The configuration is done
+ * using MicroProfile Config. The following snippet gives an example for a hypothetical Kafka connector:
+ *
+ * <pre>
+ * mp.messaging.outgoing.my-stream.type=i.e.m.reactive.messaging.impl.kafka.Kafka
+ * mp.messaging.outgoing.my-stream.bootstrap-servers=localhost:9092
+ * mp.messaging.outgoing.my-stream.topic=my-topic
+ * ...
+ * </pre>
+ * <p>
+ * This configuration keys are structured as follows: {@code mp.messaging.[incoming|outgoing].stream-name.attribute}.
```

Comment:
"Reactive messaging keys for OutgoingConnectionFactories are structured as follows: {@code mp.messaging.outgoing.stream-name.attribute}. "

In this context the [incoming/outgoing] will all be 'outgoing'? All??

If there is a read/write connector with some properties relavent to both
reading and writing are we asking user to duplicate these keys
once for incoming and once for outgoing? This seems less than
brilliant.

Also what is a user has LOTs of streams can we not have some
format that is for ALL streams of a certain type? To prevent
duplication?

Reply:
You generally don't have the same channel name for incoming and outgoing. 

If you need something shared, you can use MP Config directly as the connector is a bean, it can retrieve the configuration. For instance, the Kafka connector could contain:

```
@Inject @ConfigProperty(name="bootstrap.servers") String servers; 
```

Reply:
but would you not have to set this property with knowledge of what all the stream names were going to be? I don' think this scales well and may make it hard to seperate the concerns/config into application concerns (queue names) and devops concerns (rotating credentials etc)?

Reply:
The queue names are definitely in the `mp.messaging....` properties. The rest depends on the connector implementation. BTW, it's very likely that each channel uses a different queue name.

Reply:
I think it would be highly useful to have properties that apply to a particular messaging provider. We can insert the queue name and incoming/ougoing into a synthetic property for Configs being passed to Connectors that match by the test of the key.

So for
mp.messaging.[incoming|outgoing].stream-name.attribute

If we know the stream-name is being processed by acme-kafka  MessagingProvider
then
mp.messaging.incoming.connector.acme-kafka.ticket = dsfsadf
can result in 
mp.messaging.incoming.stream-name.ticket = dsfsadf

Or we can just spec that a connector will take account of:

mp.messaging.[incoming|outgoing].[Connector-fully-qualified-classname].attribute

Reply:
This will be handled in another PR